pertpy/tools/_distances.py

```diff
@@ -0,0 +1,407 @@
+from __future__ import annotations
+
+from typing import List, Optional, Union
+
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from ot import unif
+from ot.gromov import gromov_wasserstein
+from scipy.spatial.distance import cdist
+from sklearn.metrics import pairwise_distances
+from statsmodels.stats.multitest import multipletests
+from tqdm import tqdm
+
+"""
+DEV NOTES:
+
+User interface:
+- A function to get distances between any two or more groups of cells with a
+distance metric of choice
+- get p-values for permutation test with a distance metric of choice for all or
+    some of their groups
+
+Internally:
+- There will be two types of PermutationTest classes:
+    - One that recomputes the pairwise distance matrix every time and works for
+      any metric.
+    - One that uses the precomputed pairwise distance matrix and works only for
+    metrics that can be called on the pairwise distance matrix. Will be faster.
+NOTE: This is a bit redundant, maybe merge? Make it Fast automatically
+if Metric is callable on P?
+NOTE: Maybe move the tests to a separate file? (e.g. _tests.py) Might lead to
+confusion with DE testing though...
+
+TODO:
+- Make PermutationTest allow for multiple controls (accept list of controls)
+- Inherit docstrings from parent classes
+
+- Options how to handle Metrics for PermutationTest:
+    1. (x) Have a Metric class with two function: one of them on X and Y, the other on P.
+    The latter is only used by PermutationTest internally. The first one is accessable
+    as __call__ to the user.
+    2. Have separate, second Metric class for pairwise distances, e.g. P_Edistance.
+    3. (x) Have a wrapper function that takes a metric and returns a new metric that
+    is callable on X and Y, and uses the pairwise distance matrix P internally.
+    This function will also be accessable to the user.
+I think 1+3 is the best solution.
+"""
+
+# wrapper fcts maybe useful to user (see above)
+def distance(x: np.ndarray, y: np.ndarray, metric: str) -> float:
+    if metric == "edistance":
+        return Edistance()(x, y)
+    elif metric == "wasserstein":
+        return Wasserstein()(x, y)
+
+
+def get_distances(
+    adata: AnnData,
+    groups: list[str],
+    metric: str | Metric = "edistance",
+    **kwargs,
+) -> pd.DataFrame:
+    """Get pairwise distances between groups of cells.
+
+    Args:
+        adata (AnnData): Annotated data matrix.
+        groups (List[str]): List of group names. # or None, then use all groups
+        metric (Union[str, Metric], optional): Distance metric. Defaults to 'edistance'.
+
+    Returns:
+        pd.DataFrame: Dataframe with pairwise distances.
```

Comment:
Yup, but potentially we can also write them into the AnnData object for faster plotting etc?